### DISCUSSION TITLE
Breaking change: Multi-platform .NET 8 tags no longer support Windows containers

### DISCUSSION THREAD

**Discussion by richlander** [dotnet/dotnet-docker#4549] [Announcements]
# Breaking change: Multi-platform .NET 8 tags no longer support Windows containers

[Multi-platform container images](https://docs.docker.com/build/building/multi-platform/) make it easy to use the same tags in multiple environments, leaving `docker pull` to pull content that matches the environment. Up until .NET 8, our multi-platform tags have included support for both Linux (Debian) and Windows (Nano Server). They will only reference Linux (Debian) images with .NET 8 and later. This change was made for .NET 8 Preview 3.

This change was made because Windows multi-platform tags don't work well with respect to selecting the best version. It is better to specify the Windows version you want.

This change is discussed in much more detail in [Switch multi-platform tags to Linux only](https://github.com/dotnet/dotnet-docker/issues/4492). Separately, the referenced Linux images have been [updated to Debian Bookworm](https://github.com/dotnet/dotnet-docker/issues/4322).

Pulling a .NET 8 multi-platform tag will result in the following behavior, when "Windows Containers" mode is enabled.

```bash
>docker pull mcr.microsoft.com/dotnet/nightly/runtime:8.0-preview
8.0-preview: Pulling from dotnet/nightly/runtime
no matching manifest for windows/amd64 10.0.22621 in the manifest list entries
```

Going forward, you will need to use one of the following tags, for Windows

- `mcr.microsoft.com/dotnet/aspnet:8.0-preview-nanoserver-ltsc2022`
- `mcr.microsoft.com/dotnet/aspnet:8.0-preview-nanoserver-1809`
- `mcr.microsoft.com/dotnet/aspnet:8.0-preview-windowsservercore-ltsc2022`
- `mcr.microsoft.com/dotnet/aspnet:8.0-preview-windowsservercore-ltsc2019`

Note:

- There are also version-specific tags (like `8.0.1-nanoserver-ltsc2022`).
- The `aspnet` repo is used as an example. The same tags apply to `runtime` and `sdk` repos.

Note: The `-preview` moniker is only needed for preview releases and will not be used after .NET 8 is released in November 2023.

For example, our [aspnetapp Dockerfile](https://github.com/dotnet/dotnet-docker/blob/main/samples/aspnetapp/Dockerfile) will no longer work to build Windows images, but instead a [Windows-specific Dockerfile](https://github.com/dotnet/dotnet-docker/blob/main/samples/aspnetapp/Dockerfile.nanoserver) will be needed.

You can see that there are no longer any entries for Windows in the `8.0-preview` manifest tag.

```bash
>docker manifest inspect mcr.microsoft.com/dotnet/nightly/runtime:8.0-preview
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1368,
         "digest": "sha256:05338de18839d2d816b6eccb6ba92a2b98f57463e5780fe9a00c78d8a6ee08c4",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1368,
         "digest": "sha256:7fc93e3c438aaaa4d0e1016a1dc901a2d1b8c3270ca5ded08f4f45cd74e69857",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v7"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1368,
         "digest": "sha256:b4a15d14bdb669fbb455e9150dc882894db7f9bd6af09842ea75e0081fbe8d4e",
         "platform": {
            "architecture": "arm64",
            "os": "linux",
            "variant": "v8"
         }
      }
   ]
}
```

**Comment by JensNordenbro** [comment]
Is there any good link on this "This change was made because Windows multi-platform tags don't work well with respect to selecting the best version. It is better to specify the Windows version you want." to gain further insight?

**Reply by mthalman** [reply]
This: https://github.com/containerd/containerd/issues/6508

The behavior can be demonstrated with the multi-platform tags from older versions of .NET (e.g. `6.0`). What happens when you pull `mcr.microsoft.com/dotnet/sdk:6.0` with Docker Desktop running on Windows 11? I think most people would expect that you would get the latest Windows version available that's compatible with Windows 11: a Windows Server 2022 (build 20348) container. But you actually get the oldest that's available: Windows Server 2019 (build 17763).

**Reply by JensNordenbro** [reply]
Ok, Thanks!
